politeiad/backendv2/tstorebe/plugins/pi/statusescache.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package pi
+
+import (
+	"container/list"
+	"sync"
+
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/plugins/pi"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+)
+
+// statusEntry represents a cached proposal status and proposal data required
+// to determine the proposal status.
+type statusEntry struct {
+	propStatus pi.PropStatusT
+
+	// The following fields cache data in order to reduce the number of backend
+	// calls required to determine the proposal status.
+	recordState          backend.StateT
+	recordStatus         backend.StatusT
+	voteStatus           ticketvote.VoteStatusT
+	voteMetadata         *ticketvote.VoteMetadata
+	billingStatusesCount int // Number of billing status changes
+}
+
+// statusesCacheLimit is the cache's default maximum capacity. Note that it's
+// a var in order to allow setting different limit values in test files.
+var statusesCacheLimit = 1000
+
+// proposalStatuses is used to cache proposal data required to determine
+// the proposal status at runtime such as record metadata, vote metadata, the
+// vote status and the proposal billing status changes. The cache is necessary
+// to improve the performance and to reduce the number of backend calls when
```

Comment:
I would be a little more specific here.  It says backend calls, but it's not all backend calls that we're worried about.  It's backend calls that result in the tlog tree be retrieved, which gets very expensive when a tree contains tens of thousands of ticket vote leaves.